libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -210,6 +312,17 @@ public OAuthException(string body, Exception inner)
             }
         }
 
+        private bool TokenExpired(OAuthResponse token)
+        {
+            return DateTime.UtcNow > token.expiration_time;
+        }
+
+        private bool TokenOld(OAuthResponse token)
+        {
+            var halfwayExpiration = (token.expiration_time - TimeSpan.FromSeconds(token.expires_in / 2));
```

Comment:
This timespan is in conflict with the actual one which has the 5 minute delta. The interaction model between them will probably result in unexpected results.

Reply:
I added a comment explaining why this is useful.  This doesn't have to be accurate as it's an indicator to try harder to getting a new token. 

---
In reply to: [198678395](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198678395) [](ancestors = 198678395)